Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.cpp

```diff
@@ -0,0 +1,289 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "SkiaReplayCanvas.h"
+
+#if USE(COORDINATED_GRAPHICS) && USE(SKIA)
+#include "GLFence.h"
+#include "PlatformDisplay.h"
+
+WTF_IGNORE_WARNINGS_IN_THIRD_PARTY_CODE_BEGIN
+#include <skia/gpu/ganesh/GrBackendSurface.h>
+#include <skia/gpu/ganesh/SkImageGanesh.h>
+WTF_IGNORE_WARNINGS_IN_THIRD_PARTY_CODE_END
+
+namespace WebCore {
+
+WTF_MAKE_TZONE_ALLOCATED_IMPL(SkiaReplayCanvas);
+
+SkiaReplayCanvas::SkiaReplayCanvas(const IntSize& size, SkiaImageToFenceMap&& imageToFenceMap)
+    : SkNWayCanvas(size.width(), size.height())
+    , m_imageToFenceMap(WTFMove(imageToFenceMap))
+{
+}
+
+SkiaReplayCanvas::~SkiaReplayCanvas() = default;
+
+Ref<SkiaReplayCanvas> SkiaReplayCanvas::create(const IntSize& size, SkiaImageToFenceMap&& imageToFenceMap)
+{
+    return adoptRef(*new SkiaReplayCanvas(size, WTFMove(imageToFenceMap)));
+}
+
+sk_sp<SkImage> SkiaReplayCanvas::waitForRenderingCompletionAndRewrapImageIfNeeded(const SkImage* image)
+{
+    if (!image || !image->isTextureBacked())
+        return nullptr;
+
+    auto* glContext = PlatformDisplay::sharedDisplay().skiaGLContext();
+    if (!glContext || !glContext->makeContextCurrent())
+        return nullptr;
+
+    if (auto fence = m_imageToFenceMap.take(image))
+        fence->serverWait();
+
+    auto* grContext = PlatformDisplay::sharedDisplay().skiaGrContext();
+    if (image->isValid(grContext))
+        return nullptr;
+
+    // FIXME: Add error reporting mechanism, a failure from GetBackendTextureFromImage() should be visible / reported.
+    GrBackendTexture backendTexture;
+    if (!SkImages::GetBackendTextureFromImage(image, &backendTexture, false))
+        return nullptr;
+
+    return SkImages::BorrowTextureFrom(grContext, backendTexture, kTopLeft_GrSurfaceOrigin, image->colorType(), image->alphaType(), image->refColorSpace());
+}
+
+void SkiaReplayCanvas::invokeDrawFunctionWithImage(const SkImage* image, Function<void(const SkImage*)>&& drawFunction)
+{
+    if (auto wrappedImage = waitForRenderingCompletionAndRewrapImageIfNeeded(image)) {
+        drawFunction(wrappedImage.get());
+        return;
+    }
+
+    drawFunction(image);
+}
+
+void SkiaReplayCanvas::invokeDrawFunctionWithPaint(const SkPaint& paint, Function<void(const SkPaint&)>&& drawFunction)
+{
+    auto* shader = paint.getShader();
+    if (!shader) {
```

Comment:
We can also save this branch, if we use: `auto *image = shader ? shader->isAImage(&localMatrix, mode) : nullptr` below.